File: goldencheck/baseline/correlation.py
"""Correlation analyzer — Pearson (numeric-numeric) and Cramer's V (categorical-categorical)."""
from __future__ import annotations

import itertools
from typing import TYPE_CHECKING

try:
    import numpy as np
    from scipy.stats import chi2_contingency, pearsonr
except ImportError as _err:  # pragma: no cover
    raise ImportError(
        "scipy and numpy are required for deep-profiling baseline. "
        "Install them with: pip install 'goldencheck[baseline]'"
    ) from _err

import polars as pl

from goldencheck.baseline.models import CorrelationEntry

if TYPE_CHECKING:
    pass

__all__ = ["analyze_correlations", "_cramers_v"]

# Threshold constants
_STRONG_THRESHOLD = 0.7
_MODERATE_THRESHOLD = 0.4

# Maximum number of column pairs to evaluate (prevents O(n²) blowup on wide datasets)
_MAX_PAIRS = 500

# Minimum non-null rows required per pair
_MIN_ROWS = 30

# Maximum unique values for a string column to be treated as categorical
_MAX_CAT_UNIQUE = 100


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _strength(value: float) -> str:
    """Classify absolute correlation value into 'strong', 'moderate', or 'weak'."""
    abs_val = abs(value)
    if abs_val >= _STRONG_THRESHOLD:
        return "strong"
    if abs_val >= _MODERATE_THRESHOLD:
        return "moderate"
    return "weak"


def _pearson_entry(df: pl.DataFrame, col_a: str, col_b: str) -> CorrelationEntry | None:
    """Compute Pearson correlation between two numeric columns, or return None."""
    sub = df.select([col_a, col_b]).drop_nulls()
    if len(sub) < _MIN_ROWS:
        return None

    a_vals = sub[col_a].to_numpy()
    b_vals = sub[col_b].to_numpy()

    # Guard against zero-variance columns (pearsonr would raise or return NaN)
    if np.std(a_vals) == 0.0 or np.std(b_vals) == 0.0:
        return None

    corr, _ = pearsonr(a_vals, b_vals)
    if not np.isfinite(corr):
        return None

    strength = _strength(corr)
    if strength == "weak":
        return None

    return CorrelationEntry(
        columns=sorted([col_a, col_b]),
        measure="pearson",
        value=round(float(corr), 6),
        strength=strength,
        note=None,
    )


def _cramers_v(df: pl.DataFrame, col_a: str, col_b: str) -> CorrelationEntry | None:
    """Compute Cramer's V between two categorical columns, or return None.

    This function is exposed at module level for use by drift detectors.
    """
    sub = df.select([col_a, col_b]).drop_nulls()
    if len(sub) < _MIN_ROWS:
        return None

    # Build contingency table via group_by + pivot
    try:
        contingency = (
            sub.group_by([col_a, col_b])
            .agg(pl.len().alias("_cnt"))
            .pivot(on=col_b, index=col_a, values="_cnt")
            .fill_null(0)
        )
    except Exception:
        return None

    # Extract numeric matrix (drop the index column)
    index_col = col_a
    value_cols = [c for c in contingency.columns if c != index_col]
    if not value_cols:
        return None

    matrix = contingency.select(value_cols).to_numpy()
    if matrix.shape[0] < 2 or matrix.shape[1] < 2:
        return None

    try:
        chi2, _, _, _ = chi2_contingency(matrix)
    except Exception:
        return None

    n = int(matrix.sum())
    if n == 0:
        return None

    # Cramer's V formula with bias correction (Bergsma & Wicher, 2013)
    r, k = matrix.shape
    phi2 = chi2 / n
    phi2_corr = max(0.0, phi2 - (k - 1) * (r - 1) / (n - 1))
    r_corr = r - (r - 1) ** 2 / (n - 1)
    k_corr = k - (k - 1) ** 2 / (n - 1)
    denom = min(k_corr - 1, r_corr - 1)
    if denom <= 0:
        return None

    v = float(np.sqrt(phi2_corr / denom))
    if not np.isfinite(v):
        return None

    strength = _strength(v)
    if strength == "weak":
        return None

    return CorrelationEntry(
        columns=sorted([col_a, col_b]),
        measure="cramers_v",
        value=round(v, 6),
        strength=strength,
        note=None,
    )


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def analyze_correlations(df: pl.DataFrame) -> list[CorrelationEntry]:
    """Analyze pairwise correlations in *df* and return reportable entries.

    - Numeric pairs: Pearson correlation via scipy.stats.pearsonr.
    - Categorical pairs: Cramer's V via chi2_contingency.
    - Only moderate (>=0.4) or strong (>=0.7) correlations are reported.
    - At most _MAX_PAIRS column pairs are evaluated to prevent O(n²) blowup.
    - String columns must have n_unique < _MAX_CAT_UNIQUE to qualify as categorical.
    - Each pair must have at least _MIN_ROWS non-null rows.
    """
    # Partition columns by type
    numeric_cols: list[str] = [
        c for c in df.columns if df[c].dtype in (pl.Float32, pl.Float64, pl.Int8, pl.Int16,
                                                  pl.Int32, pl.Int64, pl.UInt8, pl.UInt16,
                                                  pl.UInt32, pl.UInt64)
    ]
    categorical_cols: list[str] = [
        c for c in df.columns
        if df[c].dtype in (pl.Utf8, pl.String, pl.Categorical)
        and df[c].n_unique() < _MAX_CAT_UNIQUE
    ]

    results: list[CorrelationEntry] = []

    # --- Numeric-numeric pairs ---
    num_pairs = list(itertools.combinations(numeric_cols, 2))
    for col_a, col_b in num_pairs[:_MAX_PAIRS]:
        entry = _pearson_entry(df, col_a, col_b)
        if entry is not None:
            results.append(entry)

    # --- Categorical-categorical pairs ---
    remaining_budget = _MAX_PAIRS - len(num_pairs)
    if remaining_budget > 0:
        cat_pairs = list(itertools.combinations(categorical_cols, 2))
        for col_a, col_b in cat_pairs[:remaining_budget]:
            entry = _cramers_v(df, col_a, col_b)
            if entry is not None:
                results.append(entry)

    return results

File: packages/goldencheck-js/src/core/baseline/correlation.ts
/**
 * Correlation analyzer — Pearson (numeric-numeric) and Cramer's V (categorical-categorical).
 * TypeScript port of goldencheck/baseline/correlation.py.
 * Edge-safe: uses only stats.ts utilities, no external dependencies.
 */

import type { TabularData } from "../data.js";
import type { CorrelationEntry } from "./models.js";
import { pearson, cramersV } from "../stats.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Threshold for "strong" correlation. */
const STRONG_THRESHOLD = 0.7;

/** Threshold for "moderate" correlation. */
const MODERATE_THRESHOLD = 0.4;

/** Maximum number of column pairs to evaluate. */
const MAX_PAIRS = 500;

/** Minimum non-null rows required per pair. */
const MIN_ROWS = 30;

/** Maximum unique values for a string column to be treated as categorical. */
const MAX_CAT_UNIQUE = 100;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function strength(value: number): "strong" | "moderate" | "weak" {
  const abs = Math.abs(value);
  if (abs >= STRONG_THRESHOLD) return "strong";
  if (abs >= MODERATE_THRESHOLD) return "moderate";
  return "weak";
}

function pearsonEntry(
  data: TabularData,
  colA: string,
  colB: string,
): CorrelationEntry | null {
  // Get paired non-null numeric values
  const aValues: number[] = [];
  const bValues: number[] = [];

  for (const row of data.rows) {
    const aRaw = row[colA];
    const bRaw = row[colB];
    if (aRaw === null || aRaw === undefined || bRaw === null || bRaw === undefined) continue;
    const a = Number(aRaw);
    const b = Number(bRaw);
    if (!Number.isFinite(a) || !Number.isFinite(b)) continue;
    aValues.push(a);
    bValues.push(b);
  }

  if (aValues.length < MIN_ROWS) return null;

  const corr = pearson(aValues, bValues);
  if (corr === null) return null;

  const s = strength(corr);
  if (s === "weak") return null;

  return {
    col1: colA < colB ? colA : colB,
    col2: colA < colB ? colB : colA,
    method: "pearson",
    value: Math.round(corr * 1e6) / 1e6,
    strength: s,
  };
}

function cramersVEntry(
  data: TabularData,
  colA: string,
  colB: string,
): CorrelationEntry | null {
  // Build contingency table
  const contingency = new Map<string, Map<string, number>>();
  let validCount = 0;

  for (const row of data.rows) {
    const aRaw = row[colA];
    const bRaw = row[colB];
    if (aRaw === null || aRaw === undefined || bRaw === null || bRaw === undefined) continue;
    const a = String(aRaw);
    const b = String(bRaw);

    let bMap = contingency.get(a);
    if (!bMap) {
      bMap = new Map<string, number>();
      contingency.set(a, bMap);
    }
    bMap.set(b, (bMap.get(b) ?? 0) + 1);
    validCount++;
  }

  if (validCount < MIN_ROWS) return null;

  const v = cramersV(contingency);
  if (v === null) return null;

  const s = strength(v);
  if (s === "weak") return null;

  return {
    col1: colA < colB ? colA : colB,
    col2: colA < colB ? colB : colA,
    method: "cramers_v",
    value: Math.round(v * 1e6) / 1e6,
    strength: s,
  };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Analyze pairwise correlations in the data and return reportable entries.
 *
 * - Numeric pairs: Pearson correlation.
 * - Categorical pairs: Cramer's V.
 * - Only moderate (>= 0.4) or strong (>= 0.7) correlations are reported.
 * - At most MAX_PAIRS (500) column pairs are evaluated.
 * - String columns must have nUnique < MAX_CAT_UNIQUE (100) to qualify as categorical.
 * - Each pair must have at least MIN_ROWS (30) non-null rows.
 */
export function analyzeCorrelations(data: TabularData): CorrelationEntry[] {
  // Partition columns by type
  const numericCols: string[] = [];
  const categoricalCols: string[] = [];

  for (const col of data.columns) {
    const dt = data.dtype(col);
    if (dt === "integer" || dt === "float") {
      numericCols.push(col);
    } else if (dt === "string") {
      if (data.nUnique(col) < MAX_CAT_UNIQUE) {
        categoricalCols.push(col);
      }
    }
  }

  const results: CorrelationEntry[] = [];
  let pairsEvaluated = 0;

  // --- Numeric-numeric pairs ---
  for (let i = 0; i < numericCols.length && pairsEvaluated < MAX_PAIRS; i++) {
    for (let j = i + 1; j < numericCols.length && pairsEvaluated < MAX_PAIRS; j++) {
      pairsEvaluated++;
      const entry = pearsonEntry(data, numericCols[i]!, numericCols[j]!);
      if (entry !== null) {
        results.push(entry);
      }
    }
  }

  // --- Categorical-categorical pairs ---
  for (let i = 0; i < categoricalCols.length && pairsEvaluated < MAX_PAIRS; i++) {
    for (let j = i + 1; j < categoricalCols.length && pairsEvaluated < MAX_PAIRS; j++) {
      pairsEvaluated++;
      const entry = cramersVEntry(data, categoricalCols[i]!, categoricalCols[j]!);
      if (entry !== null) {
        results.push(entry);
      }
    }
  }

  return results;
}
